tests/test_config.py

```diff
@@ -701,54 +703,132 @@ def test_json_encoders_type_adapter() -> None:
+):
+    config = ConfigDict(defer_build=True)
+    if defer_build_mode is not None:
+        config['_defer_build_mode'] = defer_build_mode
+
+    class MyModel(BaseModel):
+        model_config = config
         x: int
 
-    assert isinstance(MyModel.__pydantic_validator__, MockValSer)
-    assert isinstance(MyModel.__pydantic_serializer__, MockValSer)
+    if defer_build_mode is None or 'model' in defer_build_mode:
+        assert isinstance(MyModel.__pydantic_validator__, MockValSer)
+        assert isinstance(MyModel.__pydantic_serializer__, MockValSer)
+        assert generate_schema_calls.count == 0, 'Should respect _defer_build_mode'
+    else:
+        assert isinstance(MyModel.__pydantic_validator__, SchemaValidator)
+        assert isinstance(MyModel.__pydantic_serializer__, SchemaSerializer)
+        assert generate_schema_calls.count == 1, 'Should respect _defer_build_mode'
 
     m = MyModel(x=1)
     assert m.x == 1
+    assert m.model_dump()['x'] == 1
+    assert m.model_validate({'x': 2}).x == 2
+    assert m.model_json_schema()['type'] == 'object'
 
     assert isinstance(MyModel.__pydantic_validator__, SchemaValidator)
     assert isinstance(MyModel.__pydantic_serializer__, SchemaSerializer)
+    assert generate_schema_calls.count == 1, 'Should not build duplicated core schemas'
 
 
-def test_config_type_adapter_defer_build():
-    class MyModel(BaseModel, defer_build=True):
+@pytest.mark.parametrize('defer_build_mode', [None, tuple(), ('model',), ('type_adapter',), ('model', 'type_adapter')])
+def test_config_model_type_adapter_defer_build(
+    defer_build_mode: Optional[Tuple[Literal['model', 'type_adapter'], ...]], generate_schema_calls: CallCounter
+):
+    config = ConfigDict(defer_build=True)
+    if defer_build_mode is not None:
+        config['_defer_build_mode'] = defer_build_mode
+
+    class MyModel(BaseModel):
+        model_config = config
         x: int
 
+    is_deferred = defer_build_mode is None or 'model' in defer_build_mode
+    assert generate_schema_calls.count == (0 if is_deferred else 1)
+    generate_schema_calls.reset()
+
     ta = TypeAdapter(MyModel)
 
-    assert isinstance(ta.validator, MockValSer)
-    assert isinstance(ta.serializer, MockValSer)
+    assert generate_schema_calls.count == 0, 'Should use model generated schema'
 
-    m = ta.validate_python({'x': 1})
-    assert m.x == 1
-    m2 = ta.validate_python({'x': 2})
-    assert m2.x == 2
+    assert ta.validate_python({'x': 1}).x == 1
+    assert ta.validate_python({'x': 2}).x == 2
+    assert ta.dump_python(MyModel.model_construct(x=1))['x'] == 1
+    assert ta.json_schema()['type'] == 'object'
 
-    # in the future, can reassign said validators to the TypeAdapter
-    assert isinstance(MyModel.__pydantic_validator__, SchemaValidator)
-    assert isinstance(MyModel.__pydantic_serializer__, SchemaSerializer)
+    assert generate_schema_calls.count == (1 if is_deferred else 0), 'Should not build duplicate core schemas'
```

Comment:
@sydney-runkle was this a "bug" as previously (also in `main`) this assertion would have failed. It showed how `TypeAdapter` actually built core schema duplicated. First one was built via `TypeAdapter.__init__` and second one via `TypeAdapter.validator` using `MockValSer` of the model. So this happened when the model used `defer_build=True`.

I now fixed (attempt) it [here](https://github.com/pydantic/pydantic/pull/8939/commits/975eb228371784acd3c14b70a46c3d2215123dc5) as I noticed the strange duplicated core schema building in the test I added. The test assertion failed with the duplication but its now fixed.

This probably explains why we saw even worse performance when trying out original `defer_build` as it actually caused even more work to happen in the service. So the heavy core schemas being built 2x with `defer_build` in `TypeAdapter`.

Reply:
@MarkusSintonen,

Ah, interesting. Reviewing your related changes now.